subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTaskOperationListener.java

```diff
@@ -92,9 +94,9 @@ public void finished(BuildOperationDescriptor buildOperation, OperationFinishEve
         }
 
         if (buildOperation.getDetails() instanceof ExecuteTaskBuildOperationDetails) {
-            Task task = ((ExecuteTaskBuildOperationDetails) buildOperation.getDetails()).getTask();
-            TaskInternal taskInternal = (TaskInternal) task;
-            eventConsumer.dispatch(new DefaultTaskFinishedProgressEvent(finishEvent.getEndTime(), toTaskDescriptor(buildOperation, taskInternal), toTaskResult(taskInternal, finishEvent)));
+            TaskInternal task = ((ExecuteTaskBuildOperationDetails) buildOperation.getDetails()).getTask();
+            InternalTaskResult result = decoratorFactory.decorate(toTaskResult(task, finishEvent), buildOperation.getId());
```

Comment:
Why are we only doing this for tasks?

Reply:
We only need it for tasks right now and the `decorate()` method only accepts `InternalTaskResult` parameters. I didn't want to add unused methods to the `OperationResultDecoratorFactory`. Or would you like to make the interface more generic and accept `InternalOperationResult` instead of `InternalTaskResult`?

Reply:
Yeah that's what I was thinking. The decorators can then check for types they care about. I assume you'll need that in the upcoming steps anyway, so we can also do it there.